### ISSUE TITLE
Too much samplers: iTowns doesn't work on Linux|Firefox|Mesa3D

### ISSUE THREAD

**Issue by peppsac** [iTowns/itowns#86]
Since d4d101af2f871dcc67070d89f7abff13928923ec the sampler count has been increased to 16.
This breaks iTowns when using Linux|Firefox|Mesa3D:

`
THREE.WebGLProgram: shader error:  0 gl.VALIDATE_STATUS false gl.getProgramInfoLog Programs with more than 16 samplers are disallowed on Mesa drivers to avoid crashing.
`

I suggest reverting this commit since it seems to be only for debug (according to the commit message: `debug slot texture`).

**Comment by gchoqueux** [comment]
I will fix,  I will add a test capabilities on max slot texture

**Comment by gchoqueux** [comment]
try this commit 37df30d32d643e6b2013a3e3247c8930280793b2

**Comment by tbroyer** [comment]
> I suggest reverting this commit since it seems to be only for debug (according to the commit message: `debug slot texture`).

FWIW: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

**Comment by peppsac** [comment]
> FWIW: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

+1
See also : http://chris.beams.io/posts/git-commit/
And a good template for git commits : http://codeinthehole.com/writing/a-useful-template-for-commit-messages/

**Comment by gchoqueux** [comment]
sorry wrong message, it's not for debug. A better message : `fix problem with slot texture`

**Comment by peppsac** [comment]
That's why code reviews are for: fixing those before they end up in git history :-)

> try this commit 37df30d

I didn't get a chance to test your fix yet.
I noticed you already pushed it to master, maybe you could have opened a PR for this one so I could have a chance to test/review it before merging?

**Comment by peppsac** [comment]
Commit 37df30d didn't help as `maxTexturesUnits` is 32.

16 is a hard-coded limit to mitigate a 2012 security issue in mesa3D (see https://dxr.mozilla.org/mozilla-central/source/dom/canvas/WebGLProgram.cpp#930 and https://bugzilla.mozilla.org/show_bug.cgi?id=777028).

So for the time being we need to limit our sampler count to be <= 16 and it already seems quite high.
Why do you want to increase it?

**Comment by peppsac** [comment]
Still not working (tested on `dev` branch, commit bf120bb67e169cb54999a65c56f33a15d726dd3f).
FYI:
- `UNMASKED_VENDOR_WEBGL` is `Intel Open Source Technology Center`
-  `UNMASKED_RENDERER_WEBGL` is `Mesa DRI Intel(R) HD Graphics 520 (Skylake GT2)`

So you might want to change the test here: https://github.com/iTowns/itowns2/commit/bf120bb67e169cb54999a65c56f33a15d726dd3f#diff-d8f455b05139f20d311e334a269d4dbbR271

Anyway, after fixing the test there's another issue:

```
THREE.WebGLShader: Shader couldn't compile.
THREE.WebGLShader: gl.getShaderInfoLog() fragment ERROR: 0:30: '[' :  array index out of range '15'
```

Probably caused by the hard-coded `16` value here:  https://github.com/iTowns/itowns2/commit/b09f36d8#diff-f5c5a4b625fe789476515783a6e21153R72

Let me know if you want me to fix this, might be easier since I can test easily.

**Comment by gchoqueux** [comment]
`customFS += getColorAtIdUv(16)` should be `customFS += getColorAtIdUv(maxTexturesUnits-1)`
but with `maxTexturesUnits=32` the shader building failed with this error  `memory exhausted`.
I try another fix, and if it doesn't work, you can fix it.

**Comment by gchoqueux** [comment]
i fix only 

```
THREE.WebGLShader: Shader couldn't compile.
THREE.WebGLShader: gl.getShaderInfoLog() fragment ERROR: 0:30: '[' :  array index out of range '15'
```

You can to fix  `UNMASKED_VENDOR_WEBGL` and `UNMASKED_RENDERER_WEBGL`

**Comment by nosy-b** [comment]
Last Master [https://github.com/iTowns/itowns2/commit/37df30d32d643e6b2013a3e3247c8930280793b2](https://github.com/iTowns/itowns2/commit/37df30d32d643e6b2013a3e3247c8930280793b2) doesn't work on configuration Win + Angle because shader _GlobeFS.glsl_ tries to parse up to 16 textures
 `(else if (id == 15) return texture2D(dTextures[15],  pitUV(uv,pitScale_L01[15]));` 
But in _LayerMaterial.js_ is defined: 
`var maxTexturesUnits =  gfxEngine().glParams.maxTexturesUnits;`
which is 16 like on most configuration and **it is then decremented**, ending at 15.
It works if we keep maxTexturesUnits for pit and texture arrays.
But it shows a warning from _ThreeJS_: _WebGLRenderer: trying to use 16 texture units while this GPU supports only 16_

**Comment by nosy-b** [comment]
Ok I commited this temp correction because we can't leave a master that crashes on 80% of users

**Comment by gchoqueux** [comment]
fixed in #583
